### DISCUSSION TITLE
sam invoke local - should AWS_SESSION_TOKEN be an empty string?

### DISCUSSION THREAD

**Discussion by leepowellnbs** [aws/aws-sam-cli#6069] [Questions]
Obviously if I run my lambda within AWS the AWS_SESSION_TOKEN is populated and can be used for things such as accessing secrets. However when I run `sam invoke local` the token is just an empty string and I'm unable to use it for any of these services. 

Is this expected behaviour? Am I meant to create this session token myself? I'm using the Node18 runtime.

**Comment by uldyssian-sh** [comment]
Yes — this is expected behavior when using `sam invoke local`.

When you run a Lambda in AWS, temporary credentials are injected into the runtime environment:
- AWS_ACCESS_KEY_ID
- AWS_SECRET_ACCESS_KEY
- AWS_SESSION_TOKEN

These are real STS credentials associated with the Lambda execution role.

When you run `sam invoke local`, SAM is executing your function locally (in a Docker container or directly), and it does **not** automatically assume the Lambda execution role or mint STS credentials on your behalf. As a result:

- `AWS_SESSION_TOKEN` is empty (or unset)
- You are not running with Lambda-style temporary credentials

How credentials work with sam invoke local:

SAM simply passes through whatever AWS credentials are available on your local machine via the normal AWS credential provider chain:
- environment variables
- `~/.aws/credentials`
- SSO cache
- etc.

If your local credentials are long-term credentials, there is no session token, so `AWS_SESSION_TOKEN` will be empty.

If you want a session token locally:

You must provide temporary credentials yourself, for example by:
- assuming a role (`aws sts assume-role`)
- using AWS SSO and exporting the resulting credentials
- setting AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY / AWS_SESSION_TOKEN explicitly before running `sam invoke local`

SAM will then pass those through into the local Lambda environment.

Important note:

You should not try to “fake” or generate an AWS_SESSION_TOKEN yourself. Session tokens are issued by AWS STS and must correspond to the access key/secret key pair.

Summary:

An empty AWS_SESSION_TOKEN in `sam invoke local` is expected.
SAM does not auto-assume the Lambda execution role locally.
Provide real STS credentials locally if your code requires a session token to access AWS services.

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
